contracts/factory/src/contract.rs

```diff
@@ -201,6 +204,10 @@ impl FactoryTrait for Factory {
             init_fn_args.push_back(amp.unwrap().into_val(&env));
         }
 
+        //TODO: I don't like how this relies on providing the right sequence of input args to the
+        //array of arguments. We should try using builder here.
```

Comment:
True and that's a good point, but if the provided way does not allow us to verify the types during compile time... I'm not sure how much can we do, unless you mean importing 1:1 the actual structure, filling it up and then putting those arguments into the array.
This is a good TODO to think about, but please create an issue out of that and remove the comment.